policy/aggregation_type.go

```diff
@@ -0,0 +1,274 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package policy
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/m3db/m3metrics/generated/proto/schema"
+)
+
+// Supported aggregation types.
+const (
+	Unknown AggregationType = iota
+	Last
+	Lower
+	Upper
+	Mean
+	Median
+	Count
+	Sum
+	SumSq
+	Stdev
+	P50
+	P95
+	P99
+	P999
+	P9999
+	totalAggregationTypes = iota
+)
+
+const (
+	// AggregationIDLen is the length of the AggregationID.
+	// The AggregationIDLen will be 1 when totalAggregationTypes <= 64.
+	AggregationIDLen = (totalAggregationTypes-1)/64 + 1
+
+	aggregationTypesSeparator = ","
+)
+
+var (
+	// DefaultAggregationTypes is a default list of aggregation types.
+	DefaultAggregationTypes AggregationTypes
+
+	// DefaultAggregationID is a default AggregationID.
+	DefaultAggregationID AggregationID
+
+	// ValidAggregationTypes is the list of all the valid aggregation types
+	ValidAggregationTypes = []AggregationType{
+		Last,
+		Lower,
+		Upper,
+		Mean,
+		Median,
+		Count,
+		Sum,
+		SumSq,
+		Stdev,
+		P50,
+		P95,
+		P99,
+		P999,
+		P9999,
+	}
+
+	aggregationTypeStringMap map[string]AggregationType
+)
+
+func init() {
+	aggregationTypeStringMap = make(map[string]AggregationType, totalAggregationTypes)
+	for _, aggType := range ValidAggregationTypes {
+		aggregationTypeStringMap[aggType.String()] = aggType
+	}
+}
+
+// AggregationType defines a custom aggregation function.
+type AggregationType int
+
+// NewAggregationTypeFromSchema creates an aggregation type from a schema.
+func NewAggregationTypeFromSchema(input schema.AggregationType) (AggregationType, error) {
+	aggType := AggregationType(input)
+	if !aggType.IsValid() {
+		return Unknown, fmt.Errorf("invalid aggregation type from schema: %s", input)
+	}
+	return aggType, nil
+}
+
+// IsValid checks if an AggregationType is valid.
+func (a AggregationType) IsValid() bool {
+	return a > 0 && a < totalAggregationTypes
```

Comment:
Seems we are still making assumptions about the continuity of aggregation type id here. Can we just put the list valid aggregation types in a map? e.g.,
```
validAggregationTypes = map[AggregationType]struct{}{
  Last: struct{}{},
  ...
}
```
and use it in `init` and here?